pkg/ccl/backupccl/restore_data_processor.go

```diff
@@ -563,10 +689,43 @@ func (rd *restoreDataProcessor) ConsumerClosed() {
 		return
 	}
 	rd.cancelWorkersAndWait()
+
+	rd.mem.Close(rd.Ctx())
+	if rd.mon != nil {
+		rd.mon.Stop(rd.Ctx())
+	}
 	rd.agg.Close()
 	rd.InternalClose()
 }
 
+func reserveRestoreWorkerMemory(
+	ctx context.Context,
+	settings *cluster.Settings,
+	mem *mon.BoundAccount,
+	quotaPool *quotapool.IntPool,
+) (int, func(), error) {
+	numWorkersSetting := int(numRestoreWorkers.Get(&settings.SV))
+
+	numWorkers := 0
+	for worker := 0; worker < numWorkersSetting; worker++ {
+		if err := mem.Grow(ctx, minWorkerMemReservation); err != nil {
+			if worker != 0 {
+				break // no more memory to run workers
+			}
+			return 0, nil, errors.Wrap(err, "insufficient memory available to run restore")
```

Comment:
We need this error because the one way that this memory monitoring change can hang a job is if there's not enough memory to process a single file when we wait for quota. This gives the job some room for that.

Reply:
sorry I'm probably missing something - the behavior your mentioning is what we want, to always have at least a single worker that can do some work, but with that error we will fail without making progress, no? if all mem allocations fail, don't you want to succeed anyway with at least one worker?

Reply:
No, in this block I don't want the restore to continue if there's not enough memory to be reserved for one worker.